event/system.go

```diff
@@ -0,0 +1,145 @@
+package event
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Subscriber is a limited interface of System that only allows subscribing and
+// unsubscribing.
+type Subscriber interface {
+	Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event)
+	Unsubscribe(subID uint64)
+}
+
+// System keeps track of subscribers, and allows subscribing to and emitting
+// events.
+type System struct {
+	subMx       sync.RWMutex
+	subIDCount  uint64
+	subscribers map[Type]map[uint64]chan *Event
+	eventBuffer int
+	logger      logrus.FieldLogger
+}
+
+// NewEventSystem returns a new System.
+// eventBuffer determines the size of the Event channel buffer. Events might be
+// dropped if this buffer is full and there are no event listeners, or if events
+// are emitted very quickly and the event handler goroutine is busy. It is
+// recommended to handle events in a separate goroutine to not block the
+// listener goroutine.
+func NewEventSystem(eventBuffer int, logger logrus.FieldLogger) *System {
+	return &System{
+		subscribers: make(map[Type]map[uint64]chan *Event),
+		eventBuffer: eventBuffer,
+		logger:      logger, // TODO: Log some debug events
```

Comment:
Sounds good to write a debug log when a public API method is using (`Subscribe`, `Emit`, `Unsubscribe`, etc)